hicommonwealth/commonwealth
# Added more fined grain permission gating for groups

## Link to Issue
Closes: #7867

Modelling ticket: #7904 (Doesn't close because it is not the full model)

## Description of Changes
- Adds fine grained permissioning system to the backend for Groups.
- This allow us to gate topics by group membership based on specified permissions 
```typescript
export enum PermissionEnum {
  CREATE_THREAD = 'CREATE_THREAD',
  CREATE_COMMENT = 'CREATE_COMMENT',
  CREATE_THREAD_REACTION = 'CREATE_THREAD_REACTION',
  CREATE_COMMENT_REACTION = 'CREATE_COMMENT_REACTION',
  UPDATE_POLL = 'UPDATE_POLL',
}
```

## Test Plan
- Join a community with 2 addresses, say address A and B for reference. Use dydx for a reference community.
- Make address A an admin of dydx
- Go to http://localhost:8080/dydx/members/groups/create and create a group gating some topic
- Set the CREATE_THREAD permission for this group in the GroupPermissions table:
```sql
INSERT INTO "GroupPermissions" (group_id, permissions, created_at, updated_at)
VALUES (
    (SELECT id FROM "Groups" ORDER BY created_at DESC LIMIT 1),
    '{"CREATE_THREAD"}',
    now(),
    now()
);
```

- Try to create a thread with address B, you should get:

<img width="346" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/14794654/454ab3dd-0900-4b3e-9698-9c5341ad01cf">

- Now edit the group you had created and add address B to the allowlist.
- Now try creating a thread, it should work correctly.
- Note if you comment on your own thread, or you are an admin, this will never be gated

## Deployment Plan
1. When we deploy this, we will need to manually add this permission to the database for the appropriate group for the enterprise community in the associated ticket. This will look something like:

```sql
INSERT INTO "GroupPermissions" values (${group_id}, '{"CREATE_THREAD"}', now(), now());
```

## Other Considerations
- UI to fast follow
- Full proposed permissions model to fast follow
